internal/controller/lease_controller.go

```diff
@@ -0,0 +1,159 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	jumpstarterdevv1alpha1 "github.com/jumpstarter-dev/jumpstarter-controller/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// LeaseReconciler reconciles a Lease object
+type LeaseReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=jumpstarter.dev,resources=leases,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=jumpstarter.dev,resources=leases/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=jumpstarter.dev,resources=leases/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Lease object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.4/pkg/reconcile
+func (r *LeaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var lease jumpstarterdevv1alpha1.Lease
+	if err := r.Get(ctx, req.NamespacedName, &lease); err != nil {
+		log.Error(err, "unable to fetch Lease")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Mark expired lease as ended
+	if lease.Status.EndTime != nil && time.Now().After(lease.Status.EndTime.Time) {
+		lease.Status.Ended = true
+
+		if err := r.Status().Update(ctx, &lease); err != nil {
+			log.Error(err, "unable to update Lease status")
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Requeue aquire lease on EndTime
+	if lease.Status.Exporter != nil {
+		if lease.Status.EndTime == nil {
+			err := fmt.Errorf("Lease Status missing EndTime")
+			log.Error(err, "unable to requeue reconcile")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{
+			RequeueAfter: lease.Status.EndTime.Sub(time.Now()),
+		}, nil
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(&lease.Spec.Selector)
+	if err != nil {
+		log.Error(err, "Error creating selector for label selector")
+		return ctrl.Result{}, err
+	}
+
+	// List all exporters matching selector
+	var exporters jumpstarterdevv1alpha1.ExporterList
```

Comment:
We could label leased exporters, and un-label them to avoid filter/processing the busy ones. But that's an optimization, I think we may not need it for the scale, and this will work quite well.

May be we can add a comment for the possible optimization, (btw it comes with the drawback of possible synchronization issues, or an admin labeling exporters as leased...)

We could mark them with jumpstarter.dev/leased=lease-name

Reply:
May be even if we don't use the label for filtering it could be useful to reflect the lease on the exporter (for visibility...) or even a status field?

Reply:
Added a status field referencing back to lease.